Add return_exceptions to litellm.batch_completion

ffreemt opened the pull request:
This PR adds a kwarg ``return_exceptions`` to litellm.batch_completion. When `return_exceptions` is set to False (default),  `batch_completion` behaves exactly as before. But when batch_completion is set to `True`, batch_completion will return exceptions and partial resuslt instead of throwing exceptions.

The idea of `return_exceptions` is borrowed from `asyncio.gather`. `batch_completion` makes use of `concurrent.futures.ProcessPoolExecutor to run multiple copies of `litellm.completion`.  However, `litellm.completion` may fail and throw exceptions. When this happens, the patched `batch_completion` will return partial results when `return_exceptions` is set to True. 

   ``return_exceptions`` can be useful in many scenarios.

ffreemt committed:
Add return_exceptions to litellm.batch_completion for optionally returing exceptions and partial resuslt instead of throwing exceptions

ishaan-jaff commented on litellm/main.py:
```diff
@@ -2361,7 +2362,17 @@ def chunks(lst, n):
         # Retrieve the results from the futures
-        results = [future.result() for future in completions]
+        # results = [future.result() for future in completions]
+        if return_exceptions:
```
this is an undefined variable - you're never reading it from `kwargs`

ishaan-jaff reviewed (changes requested):
please add this to docs too https://docs.litellm.ai/docs/completion/batching

ffreemt committed:
Add litellm\tests\test_batch_completion_return_exceptions.py

ffreemt committed:
Merge branch 'BerriAI:main' into main

ffreemt commented:
Somehow one line (`return_exceptions: bool = False,`) has gone missing in the shuffling. I've added the line back and also added a `test` (litellm/litellm/tests/test_batch_completion_return_exceptions.py)

I am not sure how to add docs to  https://docs.litellm.ai/docs/completion/batching. But the docs part is present in the `batch_completion` function definition.
```
return_exceptions (bool): Whether to return exceptions and partial results when exceptions occur. Defaults to False.
```

ishaan-jaff reviewed (approved):
lgtm !

ishaan-jaff commented:
This failed our CI/CD on your test @ffreemt 
CI/Cd log: https://app.circleci.com/pipelines/github/BerriAI/litellm/9278/workflows/ad7ac3db-499d-41a8-9128-0c7cdf9aa041/jobs/19795/tests#failed-test-0

```
Failed: An error occurred: OpenAIException - Error code: 400 - {'error': {'message': 'Unrecognized request argument supplied: return_exceptions', 'type': 'invalid_request_error', 'param': None, 'code': None}}
```

ishaan-jaff commented:
@ffreemt reverting this merge - happy to merge again once the errors are fixed 
After the fixes, Can you add a screenshot of batch completions working for you locally with and without setting `return_exceptions`

ffreemt commented:
There is the pytest test at https://github.com/BerriAI/litellm/pull/3397/commits/a7ec1772b1457594d3af48cdcb0a382279b841c7#diff-44852387ceb00aade916d6b314dfd5d180499e54f35209ae9c07179febe08b4b showing `batch completions are working for me locally with and without setting return_exceptions`.

I'll have a look and see if I can make that CI/CD  happy.

ishaan-jaff commented:
thanks, I appreciate it @ffreemt !
